templates/new_service/.github/workflows/build-and-deploy.yml

```diff
@@ -0,0 +1,100 @@
+name: Build and Deploy
+concurrency: build_and_deploy_${{ github.ref_name }}
+
+on:
+  push:
+   branches:
+    - main
+  pull_request:
+    branches:
+    - main
+    types: [opened, reopened, synchronize, labeled]
+
+env:
+  TERRAFORM_BASE: terraform/application
+  HEALTHCHECK_CMD: 'healthcheck/all'
+
+jobs:
+  build:
+    name: Build
+    env:
+      DOCKER_REPOSITORY: #DOCKER_REPOSITORY#
+    outputs:
+      docker-image-tag: ${{ steps.build-image.outputs.tag }}
+    runs-on: ubuntu-latest
+
+    steps:
+    - uses: actions/checkout@v4
+      name: Checkout
+
+    - name: Build and push docker image
+      id: build-image
+      uses: DFE-Digital/github-actions/build-docker-image@master
+      with:
+        github-token: ${{ secrets.GITHUB_TOKEN }}
+        context: .
+        docker-repository: ${{ env.DOCKER_REPOSITORY }}
+        max-cache: true
+        reuse-cache: true
+        snyk-token: ${{ secrets.SNYK_TOKEN }}
+
+  deploy-review-app:
+    name: Deployment To Review
+    concurrency: deploy_review_${{ github.event.pull_request.number }}
+    if: ${{ github.event_name == 'pull_request' && contains(github.event.pull_request.labels.*.name, 'deploy') }}
+    needs: [build]
+    environment:
+      name: review
+      url: ${{ steps.deploy_review.outputs.environment_url }}
+    runs-on: ubuntu-latest
+    permissions:
+      id-token: write
+      pull-requests: write
+
+    steps:
+    - name: Deploy App to Review
+      id: deploy_review
+      uses: DFE-Digital/github-actions/deploy-to-aks@master
+      with:
+        azure-credentials:  ${{ secrets.AZURE_CREDENTIALS }}
+        environment: review
+        pr-number: ${{ github.event.pull_request.number }}
+        sha: ${{ needs.build.outputs.docker-image-tag }}
+        terraform-base: ${{ env.TERRAFORM_BASE }}
+        healthcheck: ${{ env.HEALTHCHECK_CMD }}
+        db-seed: true
+        smoke-test: true
+#        gcp-wip: ${{ vars.GCP_WIP }}
+#        gcp-project-id: ${{ vars.GCP_PROJECT_ID }}
+
+  deploy:
+    name: Deploy environments
+    if: ${{ github.ref == 'refs/heads/main' && github.event_name == 'push' }}
+    environment:
+      name: ${{ matrix.environment }}
+      url: ${{ steps.deploy_app.outputs.environment_url }}
+    needs: [build]
```

Comment:
I think we should add a placeholder "test" job and make deploy depend on it. devs can then add their unit tests at the right place

Reply:
I took a look at our services, and it's more common now to run test and lint in a separate workflow on push to any branch, where build and deploy runs on pr and push to main. I can see benefits to that approach.